README.md

# 通达信客户端FLUTTER版本

目前仅支持基本行情查询，之后有时间再搞出扩展和交易接口

## 功能列表

#### 基本行情

    + [x] 查询交易品种
    + [x] 查询板块信息
    + [x] 查询k线


 ```
 import 'package:tdx/tdx.dart';


 TdxClientService clientService = new TdxClientService();
 await clientService.start();
 int count0 = await clientService.getCount(Market.sh);
 print("获取到的数量为$count0");
 int count1 = await clientService.getCount(Market.sz);
 print("获取到的数量为$count1");

 List<Quote> quotes = await clientService.getQuotes(category: Category.day,
 market: Market.sz,code: "000001",count: 100,start: 1);

 print("获取到的k线熟练为${quotes.length}");
 expect(quotes.length, 100);



 List<Quote> indexQuotes = await clientService.getIndexQuotes(category: Category.day,
     market: Market.sh,code: "000001",count: 100,start: 1);

 print("获取到的k线熟练为${indexQuotes.length}");
 expect(indexQuotes.length, 100);
 ```

CHANGELOG.md
# [0.0.1]

* 基本功能


test/tdx_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:tdx/tdx.dart';

void main() {
  test('获取产品数量', () async {
    TdxClientService clientService = new TdxClientService();
    await clientService.start();
    int count0 = await clientService.getCount(Market.sh);
    print("获取到的数量为$count0");
    int count1 = await clientService.getCount(Market.sz);
    print("获取到的数量为$count1");

    List<Quote> quotes = await clientService.getQuotes(category: Category.day,
    market: Market.sz,code: "000001",count: 100,start: 1);

    print("获取到的k线熟练为${quotes.length}");
    expect(quotes.length, 100);



    List<Quote> indexQuotes = await clientService.getIndexQuotes(category: Category.day,
        market: Market.sh,code: "000001",count: 100,start: 1);

    print("获取到的k线熟练为${indexQuotes.length}");
    expect(indexQuotes.length, 100);

  });
}


lib/src/data.dart
class IndexQuote extends Quote {
  int upCount;
  int downCount;
}

class Quote extends Data {
  String toString() {
    return "$date $open-$close";
  }

  double amt;
  double low;
  double high;
  double open;
  double close;
  int vol;

  Quote(
      {this.low,
      this.high,
      this.vol,
      this.open,
      this.close,
      this.amt,
      String date})
      : super(date: date);

  bool isUp() {
    return close >= open;
  }

  static Quote mapper(v) {
    return Quote(
        low: v['low'],
        high: v['high'],
        open: v['open'],
        close: v['close'],
        date: v['date'],
        vol: v['vol'],
        amt: v['amt']);
  }

  static double getRate(Quote f, Quote l) {
    return (l.close - f.close) / f.close;
  }
}

class Data {
  String date;

  Data({this.date});
}


lib/src/client.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:math' as Math;
import 'dart:typed_data';

import 'package:convert/convert.dart';
import 'package:tdx/src/data.dart';
import 'package:tdx/src/socket_client.dart';

part 'package:tdx/src/_commands.dart';
part 'package:tdx/src/_stream.dart';

enum Category {
  m5, //0
  m15, //1
  m30, //2
  hour, //3
  day, //4
  week, //5
  month, //6
  m1, //7
  _m1, //8
  _d, //9
  season, //10
  year, //11

}
enum Market {
  sz, //0 深圳
  sh //1 上海

}

class TdxQuotesReader {
  static Quote parseForDay(Uint8List bytes, int start) {
    int time = HexUtils.readInt(bytes, start);
    double open = HexUtils.readInt(bytes, start + 4) / 100.0;
    double high = HexUtils.readInt(bytes, start + 8) / 100.0;
    double low = HexUtils.readInt(bytes, start + 12) / 100.0;
    double close = HexUtils.readInt(bytes, start + 16) / 100.0;
    double amt = HexUtils.readInt(bytes, start + 20) / 100.0;
    int vol = HexUtils.readInt(bytes, start + 24);
    Quote quote = new Quote();
    quote.date = GetQuotesCommand.getDateYmd(
        time ~/ 10000, time % 10000 ~/ 100, time % 10000 % 100);
    quote.close = (close);
    quote.open = (open);
    quote.low = (low);
    quote.high = (high);
    quote.vol = (vol);
    quote.amt = (amt);
    return quote;
  }

  static String parseDate(ByteData bdata, int start) {
    int time = bdata.getUint16(start);
    int year = time ~/ 2048 + 2004;
    int month = time % 2048 ~/ 100;
    int date = time % 2048 % 100;
    int minute = bdata.getUint16(start + 2);
    int hour = minute ~/ 60;
    minute %= 60;
    return GetQuotesCommand.getDate(year, month, date, hour, minute);
  }

  static Quote parseForMin(Uint8List bytes, int start) {
    ByteData bdata = new ByteData.view(bytes.buffer);
    String date = parseDate(bdata, start);
    double open = bdata.getFloat32(start + 4);
    double high = bdata.getFloat32(start + 8);
    double low = bdata.getFloat32(start + 12);
    double close = bdata.getFloat32(start + 16);
    double amt = bdata.getFloat32(start + 20);
    int vol = HexUtils.readInt(bytes, start + 24);
    Quote quote = new Quote();
    quote.date = date;
    quote.close = (close);
    quote.open = (open);
    quote.low = (low);
    quote.high = (high);
    quote.vol = (vol);
    quote.amt = (amt);
    return quote;
  }
}

class TdxClient extends SocketClient {
  TdxClient(String host, int port) : super(host: host, port: port);

  @override
  Future onConnectSuccess(Socket socket) async {
    try {
      await _login();
    } catch (e) {
      await close();
    }
  }

  Future<List<IndexQuote>> getIndexQuotes(
      Category category, Market market, String code, int start, int count) {
    return lock.synchronized(() async {
      return new GetIndexQuotesCommand(
              market: market,
              category: category,
              count: count,
              start: start,
              code: code)
          .process(this);
    });
  }

  Future<int> getCount(Market market) {
    return lock.synchronized(() async {
      return new GetCountCommand(
        market: market,
      ).process(this);
    });
  }

  Future<List<Quote>> getQuotes(
      Category category, Market market, String code, int start, int count) {
    return lock.synchronized(() async {
      return new GetQuotesCommand(
              market: market,
              category: category,
              count: count,
              start: start,
              code: code)
          .process(this);
    });
  }

  List<int> _buffer;
  int _packSize;
  int _uncompressSize;

  TdxInputStream _inputStream = new TdxInputStream();

  void onData(Uint8List data) {
    print("receive data ${data.length}");
    //   print("recv data $data");
    if (_buffer == null) {
      _buffer = data;
    } else {
      if (_buffer is Uint8List) {
        List<int> buffer = new List.from(_buffer);
        _buffer = buffer;
      }
      _buffer += data;
    }
    if (_buffer.length < 0x10) {
      return;
    }
    _inputStream.setData(_buffer);
    //开始解析,此时必然有头
    _inputStream.skip(12);
    _packSize = _inputStream.readShort();
    _uncompressSize = _inputStream.readShort();
    //此后应该有packSize大小，也就是说整个包大小为packSize+0x10
    if (_buffer.length < _packSize + 0x10) {
      return;
    }

    //此时已经接受到了完整的包，需要开始解析了
    if (_packSize != _uncompressSize) {
      //解压缩
      ZLibDecoder decoder = new ZLibDecoder();
      List<int> decoded = decoder.convert(_buffer.sublist(0x10));
      _buffer = decoded;
      _inputStream.setData(_buffer);
    }

    _buffer = null;
    onComplete();
  }

  _login() async {
    await writeHex("0c0218930001030003000d0001");
    await writeHex("0c0218940001030003000d0002");
    await writeHex(
        "0c031899000120002000db0fd5d0c9ccd6a4a8af0000008fc22540130000d500c9ccbdf0d7ea00000002");
  }
}

class TdxClientService {
  static TdxClientService _instance;

  TdxClientService._();

  factory TdxClientService() {
    if (_instance == null) {
      _instance = new TdxClientService._();
    }
    return _instance;
  }

  TdxClient _client = new TdxClient("119.147.212.81", 7709);
  Timer _timer;

  Future start() async {
    //从几个ip中选择一下
    _timer = new Timer.periodic(new Duration(milliseconds: 1000), onTimer);
    return _client.connect();
  }

  void onTimer(Timer tier) {}

  Future<List<Quote>> getQuotes(
      {Category category, Market market, String code, int start, int count}) {
    return _client.getQuotes(category, market, code, start, count);
  }

  Future<List<Quote>> getIndexQuotes(
      {Category category, Market market, String code, int start, int count}) {
    return _client.getIndexQuotes(category, market, code, start, count);
  }

  Future<int> getCount(Market market) {
    return _client.getCount(market);
  }

  Future close() async {
    _timer?.cancel();
    await _client?.close();
  }
}

class HexUtils {
  /// convert string to hex bytes
  static List<int> decodeHex(String data) {
    return hex.decode(data);
  }

  /// convert hex bytes to hex string
  static String encodeHex(List<int> data) {
    return hex.encode(data);
  }

  static int readInt(List<int> data, int start) {
    return data[start + 3] << 24 & 0xff000000 |
        data[start + 2] << 16 & 0xff0000 |
        data[start + 1] << 8 & 0xff00 |
        data[start] & 0xff;
  }
}


lib/src/_stream.dart
part of 'client.dart';

class TdxOutputStream {
  List<int> _data;

  TdxOutputStream() : _data = [];

  void writeInt(int i) {
    write(i & 0xff);
    write((i >> 8) & 0xff);
    write((i >> 16) & 0xff);
    write((i >> 24) & 0xff);
  }

  void write(int i) {
    _data.add(i);
  }

  void writeAscii(String str) {
    Uint8List list = ascii.encode(str);
    _data.addAll(list);
  }

  void writeFloat(double value) {
    writeInt((value * 1000).round());
  }

  void writeShort(int i) {
    write(i & 0xff);
    write((i >> 8) & 0xff);
  }

  void clear() {
    _data.clear();
  }

  List<int> toByteArray() {
    return _data;
  }

  void writeHexString(String str) {
    _data.addAll(hex.decode(str));
  }
}

class TdxInputStream {
  List<int> _data;
  int _pos;
  int _count;

  TdxInputStream();

  bool get hasData => _pos < _count;

  void setData(List<int> data) {
    this._data = data;
    this._pos = 0;
    this._count = data.length;
  }

  int readShort() {
    int first = read();
    int second = read();
    return (first & 0xff) | ((second << 8) & 0xff00);
  }

  int read() {
    return (_pos < _count) ? (_data[_pos++] & 0xff) : -1;
  }

  void skip(int n) {
    _pos += n;
  }

  int readInt() {
    int a0 = read();
    int a1 = read();
    int a2 = read();
    int a3 = read();

    return (a0 & 0xff) |
        ((a1 << 8) & 0xff00) |
        ((a2 << 16) & 0xff0000) |
        ((a3 << 24) & 0xff000000);
  }

  double getPrice() {
    int posByte = 6;
    int bdata = readByte();
    int intdata = bdata & 0x3f;
    bool sign;
    if ((bdata & 0x40) != 0) {
      sign = true;
    } else {
      sign = false;
    }
    if ((bdata & 0x80) != 0) {
      while (true) {
        bdata = readByte();
        intdata += (bdata & 0x7f) << posByte;
        posByte += 7;

        if ((bdata & 0x80) != 0) {
          continue;
        }
        break;
      }
    }

    if (sign) {
      intdata = -intdata;
    }
    return intdata.toDouble();
  }

  int readByte() {
    return read();
  }
}


lib/src/socket_client.dart
import 'dart:async';
import 'dart:io';
import 'dart:typed_data';

import 'package:convert/convert.dart';
import 'package:synchronized/synchronized.dart';

abstract class SocketClient {
  Socket _socket;
  Lock _lock = new Lock(reentrant: true);
  String host;
  int port;
  Completer _completer;

  SocketClient({this.host, this.port});

  Lock get lock => _lock;

  Future connect() {
    assert(_socket == null);
    return _lock.synchronized(() async {
      _socket = await Socket.connect(host, port);
      print("Connect to host $host success");
      _socket.listen(onData, onDone: _onDone, onError: _onError);
      await onConnectSuccess(_socket);
    });
  }

  void onData(Uint8List data);

  Future close() async {
    return _lock.synchronized(() async {
      if (_socket != null) {
        await _socket.close();
        _socket = null;
      }
    });
  }

  _onError(e) {
    if (_completer != null) {
      if (!_completer.isCompleted) {
        _completer.completeError(e);
        _completer = null;
      }
    }
  }

  void _onDone() async {
    print("======================socket is done,try to reconnect");
    _socket = null;
    try {
      await connect();
    } catch (e) {}
  }

  void onComplete() {
    if (_completer != null) {
      if (!_completer.isCompleted) {
        _completer.complete();
        _completer = null;
      }
    }
  }

  Future onConnectSuccess(Socket socket);

  Future writeHex(String command) {
    return write(hex.decode(command));
  }

  Future write(List<int> command) async {
    if (_socket == null) {
      try {
        await connect();
      } catch (e) {
        throw new Exception("Socket 未连接");
      }
    }
    _completer = new Completer();
    _socket.add(command);
    print("write data to tdx ${command.length}");
    return _completer.future;
  }
}


lib/src/_commands.dart
part of 'client.dart';

abstract class TxdCommand<T> {
  Future<T> process(TdxClient client);
}

abstract class BaseCommand<T> implements TxdCommand<T> {
  @override
  Future<T> process(TdxClient client) async {
    TdxOutputStream out = new TdxOutputStream();
    doOutput(out);
    await client.write(out.toByteArray());
    return doInput(client._inputStream);
  }

  void doOutput(TdxOutputStream out);

  T doInput(TdxInputStream stream);
}

abstract class ListCommand<T> extends BaseCommand<List<T>> {
  @override
  List<T> doInput(TdxInputStream stream) {
    int count = stream.readShort();
    List<T> result = <T>[];
    for (int i = 0; i < count; ++i) {
      result.add(readItem(stream));
    }
    return result;
  }

  T readItem(TdxInputStream stream);
}

class GetCountCommand extends BaseCommand<int> {
  Market market;

  GetCountCommand({this.market});

  @override
  int doInput(TdxInputStream stream) {
    return stream.readShort();
  }

  @override
  void doOutput(TdxOutputStream out) {
    out.writeHexString("0c0c186c0001080008004e04");
    out.writeShort(market.index);
    out.writeHexString("75c73301");
  }
}

class GetIndexQuotesCommand extends GetQuotesCommand<IndexQuote> {
  GetIndexQuotesCommand(
      {Market market, int count, int start, Category category, String code})
      : super(
            market: market,
            count: count,
            start: start,
            category: category,
            code: code);

  @override
  Quote createQuote() {
    return new IndexQuote();
  }

  @override
  IndexQuote readItem(TdxInputStream stream) {
    IndexQuote quote = super.readItem(stream);
    int upCouont = stream.readShort();
    int downCount = stream.readShort();
    quote.upCount = upCouont;
    quote.downCount = downCount;
    return quote;
  }
}

class GetQuotesCommand<T extends Quote> extends ListCommand<T> {
  Market market;
  Category category;
  String code;
  int count;
  int start;

  GetQuotesCommand(
      {this.market, this.count, this.start, this.category, this.code}) {
    if (count > 800) {
      count = 800;
    }
  }

  double preDiffBase = 0;

  @override
  void doOutput(TdxOutputStream outputStream) {
    outputStream.writeShort(0x10c);
    outputStream.writeInt(0x01016408);
    outputStream.writeShort(0x1c);
    outputStream.writeShort(0x1c);
    outputStream.writeShort(0x052d);

    outputStream.writeShort(market.index);
    outputStream.writeAscii(code);
    outputStream.writeShort(category.index);
    outputStream.writeShort(1);
    outputStream.writeShort(start);
    outputStream.writeShort(count);

    outputStream.writeInt(0);
    outputStream.writeInt(0);
    outputStream.writeShort(0);
  }

  static String readDate(TdxInputStream inputStream, int category) {
    int year;
    int month;
    int hour;
    int minute;
    int day;
    if (category < 4 || category == 7 || category == 8) {
      int zipday = inputStream.readShort();
      int tminutes = inputStream.readShort();
      year = (zipday >> 11) + 2004;
      month = ((zipday % 2048) ~/ 100);
      day = (zipday % 2048) % 100;

      hour = (tminutes ~/ 60);
      minute = tminutes % 60;
      return getDate(year, month, day, hour, minute);
    } else {
      int zipday = inputStream.readInt();
      year = (zipday ~/ 10000);
      month = ((zipday % 10000) ~/ 100);
      day = zipday % 100;
      return getDateYmd(year, month, day);
    }
  }

  static String getDateYmd(int year, int month, int date) {
    return "${toFixed(year)}${toFixed(month)}${toFixed(date)}";
  }

  static String getDate(int year, int month, int date, int hour, int minute) {
    return "${toFixed(year)}${toFixed(month)}${toFixed(date)}${toFixed(hour)}${toFixed(minute)}";
  }

  static String toFixed(int num) {
    if (num < 10) {
      return "0$num";
    }
    return "$num";
  }

  static double getPrice(double base, double diff) {
    return (base + diff) / 1000;
  }

  Quote createQuote() {
    return new Quote();
  }

  @override
  T readItem(TdxInputStream stream) {
    String date = readDate(stream, category.index);
    double priceOpenDiff = stream.getPrice();
    double priceCloseDiff = stream.getPrice();
    double priceHighDiff = stream.getPrice();
    double priceLowDiff = stream.getPrice();

    int volRow = stream.readInt();
    int vol = getVolumn(volRow).toInt();
    int dbvolRow = stream.readInt();
    double amt = getVolumn(dbvolRow);

    double open = getPrice(priceOpenDiff, preDiffBase);
    priceOpenDiff = priceOpenDiff + preDiffBase;

    double close = getPrice(priceOpenDiff, priceCloseDiff);
    double high = getPrice(priceOpenDiff, priceHighDiff);
    double low = getPrice(priceOpenDiff, priceLowDiff);

    preDiffBase = priceOpenDiff + priceCloseDiff;
    T quote = createQuote();

    quote.date = (date);
    quote.close = (close);
    quote.open = (open);
    quote.high = (high);
    quote.low = (low);
    quote.vol = (vol);
    quote.amt = (amt);
//    quote.upCount = upCouont;
//    quote.downCount = downCount;
    return quote;
  }

  static double getVolumn(int ivol) {
    int logpoint = ivol >> (8 * 3);
    int hleax = (ivol >> (8 * 2)) & 0xff;
    int lheax = (ivol >> 8) & 0xff;
    int lleax = ivol & 0xff;

    int dwEcx = logpoint * 2 - 0x7f;
    int dwEdx = logpoint * 2 - 0x86;
    int dwEsi = logpoint * 2 - 0x8e;
    int dwEax = logpoint * 2 - 0x96;

    int tmpEax;
    if (dwEcx < 0) {
      tmpEax = -dwEcx;
    } else {
      tmpEax = dwEcx;
    }

    double dblXmm6 = 0.0;
    dblXmm6 = Math.pow(2.0, tmpEax);
    if (dwEcx < 0) {
      dblXmm6 = 1.0 / dblXmm6;
    }

    double dblXmm4 = 0;
    double tmpdblXmm3;
    if (hleax > 0x80) {
      tmpdblXmm3 = 0.0;
      int dwtmpeax = dwEdx + 1;
      tmpdblXmm3 = Math.pow(2.0, dwtmpeax);
      double dblXmm0 = Math.pow(2.0, dwEdx) * 128.0;
      dblXmm0 += (hleax & 0x7f) * tmpdblXmm3;
      dblXmm4 = dblXmm0;
    } else {
      double dblXmm0 = 0.0;
      if (dwEdx >= 0) {
        dblXmm0 = Math.pow(2.0, dwEdx) * hleax;
      } else {
        dblXmm0 = (1 / Math.pow(2.0, dwEdx)) * hleax;
        dblXmm4 = dblXmm0;
      }
    }

    double dblXmm3 = Math.pow(2.0, dwEsi) * lheax;
    double dblXmm1 = Math.pow(2.0, dwEax) * lleax;
    if ((hleax & 0x80) != 0) {
      dblXmm3 *= 2.0;
      dblXmm1 *= 2.0;
    }

    double dblRet = dblXmm6 + dblXmm4 + dblXmm3 + dblXmm1;
    return dblRet;
  }
}
